File: ruby/testcode/ldapi_013.rb
require_relative 'shared'

module Net; module LDAP
  def self.new(opts = {}); self; end
  def self.bind_as(opts = {}); true; end
end; end

# vuln-code-snippet start ruby_ldapi_bind_filter
def auth_bind(req)
  username = req.param('username')
  password = req.param('password')
  ldap = Net::LDAP.new(host: 'ldap.example.com')
  ldap.bind_as(base: 'dc=example,dc=com', filter: "(uid=#{username})", password: password) # vuln-code-snippet vuln-line ruby_ldapi_bind_filter
  BenchmarkResponse.ok('auth complete')
end
# vuln-code-snippet end ruby_ldapi_bind_filter

File: rust/testcode/ldapi_013.rs
//! CWE-90: User input tainted into multi-condition AND filter via struct initialization.

fn ldap_search(base: &str, filter: &str) -> String {
    // In production: ldap3::LdapConn::new("ldap://...").search(base, Scope::Subtree, filter, vec!["cn", "mail"])
    format!("LDAP search in {} with filter {}", base, filter)
}

struct UserQuery {
    filter: String,
}

// vuln-code-snippet start testcodeLdapi013
pub fn handle(req: &super::shared::BenchmarkRequest) -> super::shared::BenchmarkResponse {
    let user = req.param("username");
    let q = UserQuery {
        filter: format!("(&(uid={})(active=true))", user),
    };
    let result = ldap_search("dc=example,dc=com", &q.filter); // vuln-code-snippet target-line testcodeLdapi013
    super::shared::BenchmarkResponse::ok(&result)
}
// vuln-code-snippet end testcodeLdapi013
